lib/auth/auth.go

```diff
@@ -1170,6 +1170,88 @@ func (a *Server) GetMFADevices(ctx context.Context, req *proto.GetMFADevicesRequ
 	}, nil
 }
 
+// verifyMFARespAndAddDeviceWithToken validates MFA register response and on success,
+// adds the new MFA device. This method is specifically for secrets and storage keys associated with a user token.
+func (a *Server) verifyMFARespAndAddDeviceWithToken(ctx context.Context, regResp *proto.MFARegisterResponse, token types.UserToken, deviceName string) error {
+	cap, err := a.GetAuthPreference(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Validate MFARegisterResponse and upsert the new device on success.
+	var dev *types.MFADevice
+	switch resp := regResp.GetResponse().(type) {
+	case *proto.MFARegisterResponse_TOTP:
+		if cap.GetSecondFactor() == constants.SecondFactorU2F {
+			return trace.BadParameter("cluster only allows U2F for second factor")
+		}
+
+		secrets, err := a.Identity.GetUserTokenSecrets(ctx, token.GetName())
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		dev, err = services.NewTOTPDevice(deviceName, secrets.GetOTPKey(), a.clock.Now())
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		if err := a.checkTOTP(ctx, token.GetUser(), resp.TOTP.Code, dev); err != nil {
+			return trace.Wrap(err)
+		}
+
+		if err := a.UpsertMFADevice(ctx, token.GetUser(), dev); err != nil {
+			return trace.Wrap(err)
+		}
+
+	case *proto.MFARegisterResponse_U2F:
+		if cap.GetSecondFactor() == constants.SecondFactorOTP {
```

Comment:
No need to bother with this one, they can't get a proper challenge if this is true.